### PR TITLE
feat: jobs feature

### PR DESCRIPTION
**Pull request by ShashwatDadhich**
# Description

Fixes: #3122 

Currently, there are two categories of apps in Devtron - Devtron apps and Helm apps. So with the feature proposed in this document, the goal is to add a new category of app called Jobs. The Goal is to add the following features to jobs.
Allow users to create jobs.
Have the following things to create a job. 
Configuration -
            - Allows the user to add a source code.
            - Allows the user to create Job pipelines. 
Trigger job -
- Allow the user to start the build of a job pipeline.
      -     Run history
            - Show the logs, source code and artifacts of all the builds.
      -     Jobs
             - Shows the list of all the previously created jobs.

## How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

It has been tested locally as well as in a VM

## Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR requires documentation updates?
* [x] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code.
* [x] I have commented my code, particularly in hard-to-understand areas.
* [x] I have tested it for all user roles.
* [x] I have added all the required unit/api test cases.

```release-note

```

### CODE REVIEW

File: pkg/app/AppListingService.go [hunk 1]
```diff
@@ -220,6 +222,42 @@ func (impl AppListingServiceImpl) FetchAllDevtronManagedApps() ([]AppNameTypeIdC
...
+	CiPipelineIDs := GetCIPipelineIDs(jobListingContainers)
+	JobsLastSucceededOnTime, err := impl.appListingRepository.FetchJobsLastSucceededOn(CiPipelineIDs)
+	jobContainers := BuildJobListingResponse(jobListingContainers, JobsLastSucceededOnTime)
+	return jobContainers, nil
+}
+
+func (impl AppListingServiceImpl) FetchOverviewCiPipelines(jobId int) ([]*bean.JobListingContainer, error) {
+	jobCiContainers, err := impl.appListingRepository.FetchOverviewCiPipelines(jobId)
+	if err != nil {
+		impl.Logger.Errorw("error in fetching app list", "error", err)
```
Comment (kripanshdevtron) [inline pkg/app/AppListingService.go:255]: logging need to be improved

File: pkg/app/AppListingService.go [hunk 2]
```diff
@@ -220,6 +222,42 @@ func (impl AppListingServiceImpl) FetchAllDevtronManagedApps() ([]AppNameTypeIdC
...
+		AppStatuses:   fetchJobListingRequest.AppStatuses,
+	}
+	appIds, err := impl.appRepository.FetchAppIdsWithFilter(jobListingFilter)
+	if err != nil {
+		impl.Logger.Errorw("error in fetching app ids list", "error", err)
+		return []*bean.JobContainer{}, err
+	}
+	jobListingContainers, err := impl.appListingRepository.FetchJobs(appIds, jobListingFilter.AppStatuses, string(jobListingFilter.SortOrder))
+	if err != nil {
+		impl.Logger.Errorw("error in fetching app list", "error", err)
```
Comment (kripanshdevtron) [inline pkg/app/AppListingService.go:243]: logging need to be improved

File: pkg/app/AppListingService.go [hunk 3]
```diff
@@ -220,6 +222,42 @@ func (impl AppListingServiceImpl) FetchAllDevtronManagedApps() ([]AppNameTypeIdC
...
+		AppNameSearch: fetchJobListingRequest.AppNameSearch,
+		SortOrder:     fetchJobListingRequest.SortOrder,
+		SortBy:        fetchJobListingRequest.SortBy,
+		Offset:        fetchJobListingRequest.Offset,
+		Size:          fetchJobListingRequest.Size,
+		AppStatuses:   fetchJobListingRequest.AppStatuses,
+	}
+	appIds, err := impl.appRepository.FetchAppIdsWithFilter(jobListingFilter)
+	if err != nil {
+		impl.Logger.Errorw("error in fetching app ids list", "error", err)
```
Comment (kripanshdevtron) [inline pkg/app/AppListingService.go:238]: logging need to be improved

File: internal/sql/repository/helper/AppListingRepositoryQueryBuilder.go [hunk 4]
```diff
@@ -93,8 +136,33 @@ func (impl AppListingRepositoryQueryBuilder) buildAppListingSortBy(appListingFil
...
+	whereCondition := "WHERE a.active = true and a.app_store = 2 "
+
+	if len(jobListingFilter.Teams) > 0 {
+		teamIds := strings.Trim(strings.Join(strings.Fields(fmt.Sprint(jobListingFilter.Teams)), ","), "[]")
+		whereCondition = whereCondition + "and a.team_id IN (" + teamIds + ") "
+	}
+
+	if jobListingFilter.AppNameSearch != "" {
+		likeClause := "'%" + jobListingFilter.AppNameSearch + "%'"
+		whereCondition = whereCondition + "and a.app_name like " + likeClause + " "
```
Comment (kripanshdevtron) [inline internal/sql/repository/helper/AppListingRepositoryQueryBuilder.go:154]: display_name should be used here ??

File: internal/sql/repository/AppListingRepository.go [hunk 5]
```diff
@@ -80,11 +83,48 @@ func NewAppListingRepositoryImpl(Logger *zap.SugaredLogger, dbConnection *pg.DB,
...
+
+func (impl AppListingRepositoryImpl) FetchJobsLastSucceededOn(CiPipelineIDs []int) ([]*bean.CiPipelineLastSucceededTime, error) {
+	var lastSucceededTimeArray []*bean.CiPipelineLastSucceededTime
+	if len(CiPipelineIDs) == 0 {
+		return lastSucceededTimeArray, nil
+	}
+	jobsLastFinishedOnQuery := impl.appListingRepositoryQueryBuilder.JobsLastSucceededOnTimeQuery(CiPipelineIDs)
+	impl.Logger.Debugw("basic app detail query: ", jobsLastFinishedOnQuery)
+	_, appsErr := impl.dbConnection.Query(&lastSucceededTimeArray, jobsLastFinishedOnQuery)
+	impl.Logger.Debugw("basic app detail query: ", jobsLastFinishedOnQuery)
```
Comment (kripanshdevtron) [inline internal/sql/repository/AppListingRepository.go:121]: why double logging ??

File: api/restHandler/app/BuildPipelineRestHandler.go [hunk 6]
```diff
@@ -230,32 +231,62 @@ func (handler PipelineConfigRestHandlerImpl) PatchCiPipelines(w http.ResponseWri
...
+			return
+		}
 	}
-	var environmentIds []int
-	for _, pipeline := range pipelineData {
-		environmentIds = append(environmentIds, pipeline.EnvironmentId)
+	resourceName := handler.enforcerUtil.GetAppRBACName(app.AppName)
+	var ok bool
+	if patchRequest.IsJob {
+		ok = handler.enforcer.Enforce(token, casbin.ResourceGlobal, casbin.ActionCreate, resourceName)
```
Comment (kripanshdevtron) [inline api/restHandler/app/BuildPipelineRestHandler.go:254]: why do we need double RBAC for Jobs ??

File: scripts/sql/115_jobs.up.sql [hunk 7]
```diff
@@ -0,0 +1,7 @@
+ALTER TABLE app ALTER COLUMN app_store DROP DEFAULT;
+ALTER TABLE app ALTER app_store TYPE integer USING CASE WHEN app_store=true THEN 1 ELSE 0 end;
```
Comment (kripanshdevtron) [inline scripts/sql/115_jobs.up.sql:2]: change it to app_type

File: scripts/sql/115_jobs.up.sql [hunk 8]
```diff
@@ -0,0 +1,7 @@
...
+ALTER TABLE app ALTER COLUMN app_store SET DEFAULT 0;
+ALTER TABLE app ADD COLUMN display_name varchar(250);
+ALTER TABLE app ADD COLUMN description text;
+ALTER TABLE ci_artifact ADD COLUMN is_artifact_uploaded BOOLEAN NOT NULL DEFAULT FALSE;
```
Comment (kripanshdevtron) [inline scripts/sql/115_jobs.up.sql:6]: handle old artifacts

File: pkg/pipeline/PipelineBuilder.go [hunk 9]
```diff
@@ -1026,6 +1041,10 @@ func (impl PipelineBuilderImpl) CreateCiPipeline(createRequest *bean.CiConfigReq
 		BeforeDockerBuild: string(beforeByte),
 		AuditLog:          sql.AuditLog{CreatedOn: time.Now(), UpdatedOn: time.Now(), CreatedBy: createRequest.UserId, UpdatedBy: createRequest.UserId},
 	}
+	if !createRequest.IsJob {
```
Comment (kripanshdevtron) [inline pkg/pipeline/PipelineBuilder.go:1044]: use appType instead of isJob Flag

File: api/restHandler/AppListingRestHandler.go [hunk 10]
```diff
@@ -554,6 +640,10 @@ func (handler AppListingRestHandlerImpl) FetchAppStageStatus(w http.ResponseWrit
 		common.WriteJsonResp(w, err, nil, http.StatusBadRequest)
 		return
 	}
+	v := r.URL.Query()
```
Comment (kripanshdevtron) [inline api/restHandler/AppListingRestHandler.go:643]: remove extra code

File: api/restHandler/AppListingRestHandler.go [hunk 11]
```diff
@@ -571,7 +661,7 @@ func (handler AppListingRestHandlerImpl) FetchAppStageStatus(w http.ResponseWrit
 	}
 	//RBAC enforcer Ends
 
-	triggerView, err := handler.appListingService.FetchAppStageStatus(appId)
+	triggerView, err := handler.appListingService.FetchAppStageStatus(appId, isJob)
```
Comment (kripanshdevtron) [inline api/restHandler/AppListingRestHandler.go:664]: since this API is used only in case deployable Apps so pass default value as false

**Review by gireesh-devtron — DISMISSED** [review]

**Review by kripanshdevtron — APPROVED** [review]

**Pull request merged**